libraries/rush-lib/src/logic/buildCache/AdditionalFilesHasher.ts

```diff
@@ -0,0 +1,84 @@
+// See LICENSE in the project root for license information.
+import { Async, FileSystem, Import, LegacyAdapters } from '@rushstack/node-core-library';
+import crypto from 'crypto';
+import path from 'path';
+
+import type { IOptions } from 'glob';
+
+const glob: typeof import('glob') = Import.lazy('glob', require);
+
+const globAsync = (pattern: string, options: IOptions = {}): Promise<string[]> => {
+  return LegacyAdapters.convertCallbackToPromise(glob, pattern, options);
+};
+
+export class AdditionalFilesHasher {
+  public static async getFileHashedFilesMap(
+    globPatterns: string[],
+    cwd: string
+  ): Promise<Map<string, string>> {
+    const filePaths: Set<string> = await this._expandGlobPatterns(globPatterns, cwd);
+
+    return await this._calculateFileHashes(filePaths, cwd);
+  }
+
+  private static async _expandGlobPatterns(globPatterns: string[], cwd: string): Promise<Set<string>> {
+    const allMatches: Set<string> = new Set<string>();
+
+    await Async.forEachAsync(
+      globPatterns,
+      async (pattern) => {
+        const matches: string[] = await globAsync(pattern, {
+          cwd,
+          nodir: true,
+          // We want to keep path's type unchanged,
+          // i.e. if the pattern was a  relative path, then matched paths should also be relative paths
+          //      if the pattern was an absolute path, then matched paths should also be absolute paths
+          //
+          // We are doing this because these paths are going to be used to calculate a hash for the build cache and some users
+          // might choose to depend on global files (e.g. `/etc/os-release`) and some might choose to depend on local files
+          // (e.g. `../path/to/workspace/file`)
+          //
+          // In both cases we want that path to the resource would be the same on all machines,
+          // regardless of what is the current working directory.
+          //
+          // That being said, we want to keep `realpath` and `absolute` options here as false:
+          realpath: false,
+          absolute: false
+        });
+        matches.forEach((match) => allMatches.add(match));
+      },
+      { concurrency: 10 }
+    );
+
+    if (allMatches.size === 0) {
+      throw new Error(
+        `Couldn't find any files matching provided glob patterns: ["${globPatterns.join('", "')}"].`
+      );
+    }
+
+    return allMatches;
+  }
+
+  private static async _calculateFileHashes(
+    filePaths: Set<string>,
+    cwd: string
+  ): Promise<Map<string, string>> {
+    const fileHashes: Map<string, string> = new Map<string, string>();
+
+    await Async.forEachAsync(
+      filePaths,
+      async (filepath) => {
+        const fullPath: string = path.isAbsolute(filepath) ? filepath : path.join(cwd, filepath);
+        const content: string = await FileSystem.readFileAsync(fullPath);
+
+        const hashValue: string = crypto.createHash('sha1').update(content).digest('hex');
```

Comment:
Even though these files aren't tracked by Git, we would still favor using the Git hash-object functionality abstracted by the `@rushstack/package-deps-hash` package for performance and consistency.

Reply:
That makes perfect sense - I've update the code: https://github.com/microsoft/rushstack/pull/3824/commits/b99b3bca28101947a615d1625847e32d9bd02af3. 

And a few more notes: 
- Of course, after the change I've tested everything again - everything seems to work just fine;
- I've updated variable name from `cwd` to `packagePath` to align with `getGitHashForFiles` signature; 
- As one would expect, `getGitHashForFiles` generates different hash values than custom developed; `_calculateFileHashes` - so indeed, good for consistency (even though it doesn't change the end result);
- Regarding performance, I actually got interested and ran poor's man performance test - at least on my machine delegating this work to `git` resulted in about `~30% slower` execution time than reading and hashing files using previously written `_calculateFileHashes` function.  Nonetheless, I do believe that it makes sense to prioritise consistency here as the performance difference is not _that_ impactful here.


If anyone would be interested - source code of poor's man performance test:
    <details>
      <summary>AdditionalFilesHasher.test.ts</summary>
      
```ts
import { Async, FileSystem } from '@rushstack/node-core-library';
import { getGitHashForFiles } from '@rushstack/package-deps-hash';
import crypto from 'crypto';
import path from 'path';
import { performance } from 'perf_hooks';

const calculateFileHashesNode = async (
  filePaths: string[],
  projectPath: string
): Promise<Map<string, string>> => {
  const fileHashes: Map<string, string> = new Map<string, string>();

  await Async.forEachAsync(
    filePaths,
    async (filepath) => {
      const fullPath: string = path.isAbsolute(filepath) ? filepath : path.join(projectPath, filepath);
      const content: string = await FileSystem.readFileAsync(fullPath);

      const hashValue: string = crypto.createHash('sha1').update(content).digest('hex');

      fileHashes.set(filepath, hashValue);
    },
    { concurrency: 10 }
  );

  return fileHashes;
};

const calculateFileHashesGit = (filePaths: string[], projectPath: string): Map<string, string> =>
  getGitHashForFiles(filePaths, projectPath);

describe('check calculateFileHashes performance', () => {
  const filePaths = [
    '../path/to/file1.json',
    './path/to/file2.json',
    ///...
    'path/to/file20.json'
  ];
  const projectPath = '/path/to/project';

  it('calculateFileHashesNode', async () => {
    const start = performance.now();
    for (let i = 0; i < 500; i++) {
      const result = await calculateFileHashesNode(filePaths, projectPath);
    }
    const end = performance.now();
    console.log((end - start) / 1000); // on my machine the result was 14.165640457987786
  });

  it('calculateFileHashesGit', () => {
    const start = performance.now();
    for (let i = 0; i < 500; i++) {
      const result = calculateFileHashesGit(filePaths, projectPath);
    }
    const end = performance.now();
    console.log((end - start) / 1000); // on my machine the result was 20.251019375026225
  });
});
```
      
    </details>